[wip] add `jwt` action

EltonLobo07 opened the pull request.

EltonLobo07 committed:
add initial version of the `jwt` action (wip)

EltonLobo07 commented on library/src/actions/jwt/jwt.ts:
The strategy I have used to check if the string is a valid jwt is:
- check if the characters used are valid and the string is divided in three parts (using a regex)
- since the header is an object with `typ` and optional `algorithm` properties, check the header too (using JavaScript)

Something similar was done [here](https://github.com/colinhacks/zod/pull/2666) but I think the first check with the regex is still needed), do you think it's a good idea?

fabian-hiller replied:
I will review this and the `'none'` algorithm question after we solved the other questions.

EltonLobo07 commented on library/src/actions/jwt/jwt.ts:
```diff
@@ -0,0 +1,193 @@
+} from '../../types/index.ts';
+import { _addIssue } from '../../utils/index.ts';
+
+const POSSIBLE_JWT_REGEX = /^(?:[\w-]+\.){2}[\w-]+$/u;
```
Usually the regexes are stored inside `/library/src/regex.ts`. But I think the file is used to only store complete public regexes. By complete I mean if the string passes the regex's test, the string is valid. Like I stated (the comment at the top of this file), this regex is just the first part of the whole check, so its better it is a private member than in `/library/src/regex.ts`. Also, by naming it `POSSIBLE_JWT_REGEX` I meant if the string passes the test, there is a possiblity of the string being a valid jwt. If you have a better name for this regex ref, let me know. 

fabian-hiller replied:
This looks good to me!

EltonLobo07 commented on library/src/actions/jwt/jwt.ts:
```diff
@@ -0,0 +1,193 @@
+  | 'PS256'
+  | 'PS384'
+  | 'PS512'
+  | 'none';
```
The `none` algorithm has a problem mentioned [here](https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/) under the title `Meet the "None" Algorithm` but the algorithm was present in the [RFC](https://datatracker.ietf.org/doc/html/rfc7518#section-3.1). I think it we should follow the RFC as we are just checking the structure of the string. I wanted to mention it here to let you know.

EltonLobo07 commented on library/src/actions/jwt/jwt.ts:
```diff
@@ -0,0 +1,193 @@
+  | 'PS512'
+  | 'none';
+
+function isValidJwt(
```
Would you recommend inlining this function in the `requirement` function of the implementation? Or is this okay? I think this is better for readability.

fabian-hiller replied:
I think I would implement it similar to how `creditCard` is implemented.

EltonLobo07 commented on library/src/actions/jwt/jwt.ts:
```diff
@@ -0,0 +1,193 @@
+  /**
+   * The algorithm used to sign the jwt.
+   */
+  readonly algorithm: TJwtAlgorithm;
```
For some actions, there is a `requirement` property. Initially I concluded, the algorithm can be considered as a requirement but after thinking about it, I decided to add a new key because I think a `requirement` is something complete. Example: the `length` action with requirement set to `5` would mean testing the input with the requirement, nothing else. But in this case there are two steps. So, should I use the `requirement` key? If yes, what value should I set it to? What was the purpose of the `requirement` key in the issue and action objects?

fabian-hiller replied:
I have answered this at the other comment.

EltonLobo07 commented on library/src/actions/jwt/jwt.ts:
```diff
@@ -0,0 +1,193 @@
+  /**
+   * The algorithm used to sign the jwt.
+   */
+  readonly algorithm: TJwtAlgorithm;
```
Same question mentioned [here](https://github.com/EltonLobo07/valibot/pull/1/files#r1807497874)

fabian-hiller replied:
Adding the algorithm parameter as a property to `JwtAction` looks good to me, but I would not add it to `JwtIssue` as issues so far only contain common properties. In simple words, we never added additional properties to issues in the past, but we added them to the action itself to make them accessible and allow them to be serialized and dezerialized sometime in the future.

EltonLobo07 committed:
improve the initial version of the `jwt` action

Improvements:
    - remove the `algorithm` property from the `JwtIssue` interface
      so that only common properties are present in the interface
    - add the `requirement` property to `JwtAction` & `JwtIssue`. The
      property is mapped to the jwt validation function

EltonLobo07 committed:
add doc comments to everything used in jwt action file

EltonLobo07 committed:
improve type safety & impl of the action's `requirement` fn

EltonLobo07 commented:
Thank you! I think everything is clear from this point, except the strategy used.

My plan:
1. I'll add tests
2. review the changes once again (make changes if needed)
3. create a PR for the main repository
4. you can review/make changes and merge the action along with the tests
5. I'll take care of the documentation
